File: miscellaneous/sorting/quick-sort/20230218/quick-sort.js
'use strict'

const quickSort = (nums, start, end) => {
    if (nums.length <= 1) return nums;

    let left = start;
    let right = start;
    const pivot = end;

    while (right < end) {
        if (nums[right] <= nums[pivot]) {
            [nums[left], nums[right]] = [nums[right], nums[left]];
            left++;
        }

        right++;
    }

    [nums[left], nums[pivot]] = [nums[pivot], nums[left]];

    left - 1 > start && quickSort(nums, start, left - 1);
    left + 1 < end && quickSort(nums, left + 1, end);

    return nums;
}


module.exports = quickSort;

File: miscellaneous/sorting/quick-sort/20230218/quick_sort.py
from typing import List


def quick_sort(nums: List[int], start: int, end: int) -> List[int]:
    if len(nums) < 2 or end <= start:
        return nums

    left = start
    right = start
    pivot = end

    while right < end:
        if nums[right] <= nums[pivot]:
            [nums[left], nums[right]] = [nums[right], nums[left]]
            left += 1

        right += 1

    [nums[pivot], nums[left]] = [nums[left], nums[pivot]]

    left - 1 > start and quick_sort(nums, start, left - 1)
    left + 1 < end and quick_sort(nums, left + 1, end)

    return nums
